File: Python/lark_card_kit/builders/card_builder.py
"""卡片构建器，提供流畅 API 构建卡片，支持完美的类型提示。

本模块移除了 Lambda 表达式支持，只保留两种构建方式：
1. 上下文管理器（推荐用于嵌套结构）
2. 方法链 + done()（推荐用于简单结构）
"""

from __future__ import annotations

from typing import Any, Callable, Self, TypeVar

from ..models.card import Card
from ..models.elements import Element
from ..templates.template_filler import TemplateParameterFiller
from ..utils.element_finder import ElementFinder
from ..utils.template_helper import fill_template
from .card_header_builder import CardHeaderBuilder
from .card_body_builder import CardBodyBuilder
from .card_config_builder import CardConfigBuilder
from .card_link_builder import CardLinkBuilder
from .fallback_builder import FallbackBuilder

E = TypeVar("E", bound=Element)


class CardBuilder:
    """卡片构建器，提供流畅 API 构建卡片。

    提供可链式调用的方法来构建飞书卡片。

    示例:
        >>> # 方式一：上下文管理器（推荐）
        >>> with CardBuilder.create() as builder:
        ...     with builder.header() as h:
        ...         h.title("Welcome").template("blue")
        ...     with builder.body() as b:
        ...         b.plain_text("Hello, World!")
        >>> card = builder.build()
        >>>
        >>> # 方式二：方法链 + done()
        >>> builder = CardBuilder.create()
        >>> builder.header().title("Welcome").done()
        >>> builder.body().plain_text("Hello").done()
        >>> card = builder.build()
        >>>
        >>> # 方式三：构建后修改组件
        >>> builder = CardBuilder.create()
        >>> with builder.body() as b:
        ...     with b.button() as btn:
        ...         btn.text("Click").element_id("btn1")
        >>> builder.update_element("btn1", lambda e: setattr(e, "disabled", True))
        >>> card = builder.build()
    """

    def __init__(self) -> None:
        """初始化卡片构建器。"""
        self._card = Card()
        self._filler = TemplateParameterFiller()
        self._template_applied = False

    @classmethod
    def create(cls) -> Self:
        """创建新的卡片构建器。

        返回:
            新的 CardBuilder 实例。
        """
        return cls()

    def __enter__(self) -> Self:
        """进入上下文管理器。

        返回:
            自身，供上下文内使用。
        """
        return self

    def __exit__(self, *args: Any) -> None:
        """退出上下文管理器。"""
        pass

    def config(self) -> CardConfigBuilder:
        """开始配置卡片设置。

        返回:
            CardConfigBuilder 实例。
        """
        return CardConfigBuilder(self)

    def header(self) -> CardHeaderBuilder:
        """开始配置卡片头部。

        返回:
            CardHeaderBuilder 实例。
        """
        return CardHeaderBuilder(self)

    def body(self) -> CardBodyBuilder:
        """开始配置卡片主体。

        返回:
            CardBodyBuilder 实例。
        """
        return CardBodyBuilder(self)

    def card_link(self) -> CardLinkBuilder:
        """开始配置卡片链接。

        返回:
            CardLinkBuilder 实例。
        """
        return CardLinkBuilder(self)

    def fallback(self) -> FallbackBuilder:
        """开始配置降级内容。

        返回:
            FallbackBuilder 实例。
        """
        return FallbackBuilder(self)

    def set_parameter(self, key: str, value: Any) -> Self:
        """设置模板参数。

        参数:
            key: 参数键。
            value: 参数值。

        返回:
            自身，用于链式调用。
        """
        self._filler.set_parameter(key, value)
        return self

    def set_parameters(self, params: dict[str, Any]) -> Self:
        """设置多个模板参数。

        参数:
            params: 参数键值对字典。

        返回:
            自身，用于链式调用。
        """
        self._filler.set_parameters(params)
        return self

    def _set_config(self, config: Any) -> None:
        """内部方法：设置卡片配置。"""
        self._card.config = config

    def _set_header(self, header: Any) -> None:
        """内部方法：设置卡片头部。"""
        self._card.header = header

    def _set_body(self, body: Any) -> None:
        """内部方法：设置卡片主体。"""
        self._card.body = body

    def _set_card_link(self, link: Any) -> None:
        """内部方法：设置卡片链接。"""
        self._card.card_link = link

    def _set_fallback(self, fallback: Any) -> None:
        """内部方法：设置降级内容。"""
        self._card.fallback = fallback

    def find_element(self, element_id: str) -> Element | None:
        """通过 ID 查找组件。

        参数:
            element_id: 组件 ID。

        返回:
            找到的组件，如果未找到则返回 None。

        示例:
            >>> button = builder.find_element("btn1")
            >>> if button:
            ...     button.disabled = True
        """
        if self._card.body is None:
            return None
        finder = ElementFinder(self._card.body.elements)
        return finder.find_by_id(element_id)

    def find_element_as(self, element_id: str, element_type: type[E]) -> E | None:
        """通过 ID 查找组件并转换为指定类型。

        参数:
            element_id: 组件 ID。
            element_type: 期望的组件类型。

        返回:
            找到的组件（如果类型匹配），如果未找到或类型不匹配则返回 None。

        示例:
            >>> from lark_card_kit import Button
            >>> button = builder.find_element_as("btn1", Button)
            >>> if button:
            ...     button.disabled = True
        """
        if self._card.body is None:
            return None
        finder = ElementFinder(self._card.body.elements)
        return finder.find_by_id_as(element_id, element_type)

    def update_element(self, element_id: str, updater: Callable[[Element], None]) -> Self:
        """通过 ID 查找组件并修改。

        参数:
            element_id: 组件 ID。
            updater: 修改组件的函数。

        返回:
            自身，用于链式调用。

        示例:
            >>> builder.update_element("btn1", lambda e: setattr(e, "disabled", True))
        """
        if self._card.body is not None:
            finder = ElementFinder(self._card.body.elements)
            finder.update_by_id(element_id, updater)
        return self

    def update_element_as(
        self,
        element_id: str,
        element_type: type[E],
        updater: Callable[[E], None],
    ) -> Self:
        """通过 ID 查找组件（指定类型）并修改。

        参数:
            element_id: 组件 ID。
            element_type: 期望的组件类型。
            updater: 修改组件的函数。

        返回:
            自身，用于链式调用。

        示例:
            >>> from lark_card_kit import Button
            >>> builder.update_element_as("btn1", Button, lambda b: b.disabled = True)
        """
        if self._card.body is not None:
            finder = ElementFinder(self._card.body.elements)
            finder.update_by_id_as(element_id, element_type, updater)
        return self

    def get_finder(self) -> ElementFinder | None:
        """获取 ElementFinder 实例用于高级查找操作。

        返回:
            ElementFinder 实例，如果 body 为空则返回 None。

        示例:
            >>> finder = builder.get_finder()
            >>> if finder:
            ...     buttons = finder.find_by_tag("button")
            ...     finder.update_all_by_tag("button", lambda e: setattr(e, "disabled", True))
        """
        if self._card.body is None:
            return None
        return ElementFinder(self._card.body.elements)

    def build(self) -> Card:
        """构建卡片。

        返回:
            构建好的 Card 实例。
        """
        if self._filler.has_parameters() and not self._template_applied:
            fill_template(self._card, self._filler)
            self._template_applied = True

        return self._card

    def to_json(self, indent: int | None = None) -> str:
        """构建并序列化卡片为 JSON。

        参数:
            indent: 缩进空格数。

        返回:
            卡片的 JSON 字符串表示。
        """
        return self.build().to_json(indent=indent)

File: TypeScript/LarkCardKit/src/builders/CardBuilder.ts
import type { Card, CardConfig, CardHeader, CardBody, CardLink, Fallback } from '../models';
import type { Element } from '../models/elements';
import { CardConfigBuilder } from './CardConfigBuilder';
import { CardHeaderBuilder } from './CardHeaderBuilder';
import { CardBodyBuilder } from './CardBodyBuilder';
import { CardLinkBuilder } from './CardLinkBuilder';
import { FallbackBuilder } from './FallbackBuilder';
import { TemplateParameterFiller } from '../utils/TemplateParameterFiller';
import { fillTemplate } from '../utils/TemplateHelper';
import { ElementFinder } from '../utils/ElementFinder';
import { toJson } from '../utils/JsonSerializer';

/**
 * 卡片构建器
 *
 * 提供流畅 API 构建飞书卡片，支持通过 ID 查找和修改组件。
 *
 * @example
 * ```typescript
 * // 构建卡片
 * const card = CardBuilder.create()
 *   .header(h => h.title('Welcome'))
 *   .body(b => b
 *     .plainText('Hello')
 *     .button(btn => btn.text('Click').elementId('btn1')))
 *   .build();
 *
 * // 通过 ID 修改组件
 * const builder = CardBuilder.create();
 * builder.body(b => b.button(btn => btn.text('Click').elementId('btn1')));
 * builder.modifyElement('btn1', el => { el.disabled = true; });
 * const card = builder.build();
 * ```
 */
export class CardBuilder {
  private _card: Card = {
    schema: '2.0',
    body: { elements: [] }
  };
  private _filler = new TemplateParameterFiller();
  private _finder?: ElementFinder;
  private _templateApplied = false;

  config(configure: (builder: CardConfigBuilder) => void): this {
    const builder = new CardConfigBuilder();
    configure(builder);
    this._card.config = builder.build();
    return this;
  }

  header(configure: (builder: CardHeaderBuilder) => void): this {
    const builder = new CardHeaderBuilder();
    configure(builder);
    this._card.header = builder.build();
    return this;
  }

  body(configure: (builder: CardBodyBuilder) => void): this {
    const builder = new CardBodyBuilder(this._card.body);
    configure(builder);
    return this;
  }

  cardLink(configure: (builder: CardLinkBuilder) => void): this {
    const builder = new CardLinkBuilder();
    configure(builder);
    this._card.card_link = builder.build();
    return this;
  }

  cardLinkUrl(url: string): this {
    this._card.card_link = { url };
    return this;
  }

  fallback(configure: (builder: FallbackBuilder) => void): this {
    const builder = new FallbackBuilder();
    configure(builder);
    this._card.fallback = builder.build();
    return this;
  }

  fallbackText(title: string, content: string): this {
    this._card.fallback = {
      title: { tag: 'plain_text', content: title },
      content: { tag: 'plain_text', content: content }
    };
    return this;
  }

  setParameter(key: string, value: unknown): this {
    this._filler.setParameter(key, value);
    this._templateApplied = false;
    return this;
  }

  setParameters(params: Record<string, unknown>): this {
    this._filler.setParameters(params);
    this._templateApplied = false;
    return this;
  }

  private applyTemplateParameters(): void {
    if (this._templateApplied) return;
    if (this._filler.hasParameters()) {
      fillTemplate(this._card, this._filler);
    }
    this._templateApplied = true;
  }

  private getFinder(): ElementFinder {
    this._finder ??= new ElementFinder(this._card);
    return this._finder;
  }

  /**
   * 通过 ID 查找组件
   * @param elementId 组件 ID
   * @returns 找到的组件，如果未找到则返回 undefined
   *
   * @example
   * ```typescript
   * const button = builder.findElementById('btn1');
   * if (button) {
   *   button.disabled = true;
   * }
   * ```
   */
  findElementById(elementId: string): Element | undefined {
    return this.getFinder().findElementById(elementId);
  }

  /**
   * 通过 ID 查找组件并验证类型
   * @param elementId 组件 ID
   * @param expectedTag 期望的组件 tag
   * @returns 找到的组件，如果未找到或类型不匹配则返回 undefined
   *
   * @example
   * ```typescript
   * const button = builder.findElementByIdAs<Button>('btn1', 'button');
   * ```
   */
  findElementByIdAs<T extends Element>(elementId: string, expectedTag: string): T | undefined {
    return this.getFinder().findElementById<T>(elementId, expectedTag);
  }

  /**
   * 通过 ID 修改组件
   * @param elementId 组件 ID
   * @param modifier 修改函数
   * @returns 是否成功修改
   *
   * @example
   * ```typescript
   * builder.modifyElement('btn1', (el) => {
   *   el.disabled = true;
   * });
   * ```
   */
  modifyElement(elementId: string, modifier: (element: Element) => void): boolean {
    return this.getFinder().modifyElementById(elementId, modifier);
  }

  /**
   * 通过 ID 修改组件（带类型验证）
   * @param elementId 组件 ID
   * @param expectedTag 期望的组件 tag
   * @param modifier 修改函数
   * @returns 是否成功修改
   *
   * @example
   * ```typescript
   * builder.modifyElementAs<Button>('btn1', 'button', (btn) => {
   *   btn.disabled = true;
   * });
   * ```
   */
  modifyElementAs<T extends Element>(
    elementId: string,
    expectedTag: string,
    modifier: (element: T) => void
  ): boolean {
    return this.getFinder().modifyElementByIdAs(elementId, expectedTag, modifier);
  }

  /**
   * 替换指定 ID 的组件
   * @param elementId 要替换的组件 ID
   * @param newElement 新组件
   * @returns 是否成功替换
   *
   * @example
   * ```typescript
   * builder.replaceElement('text1', { tag: 'div', text: { tag: 'plain_text', content: 'New text' } });
   * ```
   */
  replaceElement(elementId: string, newElement: Element): boolean {
    return this.replaceElementRecursive(this._card.body.elements, elementId, newElement);
  }

  private replaceElementRecursive(elements: Element[], elementId: string, newElement: Element): boolean {
    for (let i = 0; i < elements.length; i++) {
      if (elements[i].element_id === elementId) {
        elements[i] = newElement;
        return true;
      }
      // Check nested elements (Div, Form, ColumnSet, etc.)
      const elem = elements[i] as Element & { elements?: Element[] };
      if ('elements' in elem && Array.isArray(elem.elements)) {
        if (this.replaceElementRecursive(elem.elements!, elementId, newElement)) {
          return true;
        }
      }
    }
    return false;
  }

  /**
   * 通过 tag 查找所有组件
   * @param tag 组件 tag
   * @returns 找到的组件列表
   */
  findAllElementsByTag(tag: string): Element[] {
    return this.getFinder().findAllElementsByTag(tag);
  }

  /**
   * 批量修改所有指定 tag 的组件
   * @param tag 组件 tag
   * @param modifier 修改函数
   * @returns 修改的组件数量
   */
  updateAllByTag(tag: string, modifier: (element: Element) => void): number {
    return this.getFinder().updateAllByTag(tag, modifier);
  }

  build(): Card {
    this.applyTemplateParameters();
    return this._card;
  }

  toJson(): string {
    this.applyTemplateParameters();
    return toJson(this._card);
  }

  toJsonIndented(): string {
    this.applyTemplateParameters();
    return toJson(this._card, true);
  }

  static create(): CardBuilder {
    return new CardBuilder();
  }
}